src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationFilter.cs

```diff
@@ -0,0 +1,171 @@
+        }
+
+        /// <summary>
+        /// Called early in the filter pipeline to confirm request is authorized.
+        /// </summary>
+        /// <param name="context">The <see cref="T:Microsoft.AspNetCore.Mvc.Filters.AuthorizationFilterContext" />.</param>
+        public async Task OnAuthorizationAsync(AuthorizationFilterContext context)
+        {
+            Guard.NotNull(context, nameof(context));
+            Guard.NotNull(context.HttpContext, nameof(context.HttpContext));
+            Guard.For<ArgumentException>(() => context.HttpContext.Connection is null, "Invalid action context given without any HTTP connection");
+            Guard.For<ArgumentException>(() => context.HttpContext.RequestServices is null, "Invalid action context given without any HTTP request services");
+
+            ILogger logger = 
+                context.HttpContext.RequestServices
+                       .GetService<ILoggerFactory>()
+                       ?.CreateLogger<CertificateAuthenticationFilter>() 
+                ?? (ILogger) NullLogger.Instance;
+
+            ISecretProvider userDefinedSecretProvider = 
+                context.HttpContext.RequestServices.GetService<ICachedSecretProvider>()
+                ?? context.HttpContext.RequestServices.GetService<ISecretProvider>();
+
+            if (userDefinedSecretProvider == null)
+            {
+                throw new KeyNotFoundException(
+                    $"No configured {nameof(ICachedSecretProvider)} or {nameof(ISecretProvider)} implementation found in the request service container. "
+                    + "Please configure such an implementation (ex. in the Startup) of your application");
+            }
+
+            X509Certificate2 clientCertificate = context.HttpContext.Connection.ClientCertificate;
+            if (clientCertificate == null)
+            {
+                logger.LogWarning(
+                    "No client certificate was specified in the HTTP request while this authentication filter "
+                    + $"requires a certificate to validate on the {String.Join(", ", _requirements.Select(item => item.requirement))}");
+                
+                context.Result = new UnauthorizedResult();
+            }
+            else if (!await IsAllowedCertificate(clientCertificate, userDefinedSecretProvider, logger))
+            {
+                context.Result = new UnauthorizedResult();
+            }
+        }
+
+        private async Task<bool> IsAllowedCertificate(
+            X509Certificate2 clientCertificate, 
+            ISecretProvider provider,
+            ILogger logger)
+        {
+            var requirementValues = await Task.WhenAll(_requirements.Select(async item =>
+            {
+                string expected = await provider.Get(item.configurationKey);
+                return (requirement: item.requirement, key: item.configurationKey, expected: expected);
+            }));
+
+            return requirementValues.All(value =>
```

Comment:
Move all of this logic to dedicated method instead?

Reply:
Hmm, I'm not sure if that would improve readability; since I see enough relation between the two functions to place them toghether in a way seperate methods/names would make things maybe not that clear.

Reply:
I think Tom suggests to move the inline lambda expression to a method, which I can agree with.